scripts/eslint/rules/js-extensions.cjs

```diff
@@ -0,0 +1,55 @@
+const { createRule } = require("./utils.cjs");
+
+/** @typedef {import("@typescript-eslint/utils").TSESTree.ImportDeclaration | import("@typescript-eslint/utils").TSESTree.ExportAllDeclaration | import("@typescript-eslint/utils").TSESTree.ExportNamedDeclaration | import("@typescript-eslint/utils").TSESTree.TSImportEqualsDeclaration} ImportOrExport */
+
+module.exports = createRule({
+    name: "js-extensions",
+    meta: {
+        docs: {
+            description: ``,
+        },
+        messages: {
+            missingJsExtension: `This relative import is missing a '.js' extension`,
+        },
+        schema: [],
+        type: "suggestion",
+        fixable: "code",
+    },
+    defaultOptions: [],
+
+    create(context) {
+        /** @type {(node: ImportOrExport) => void} */
+        const check = node => {
+            let source;
+            if (node.type === "TSImportEqualsDeclaration") {
+                const moduleReference = node.moduleReference;
+                if (
+                    moduleReference.type === "TSExternalModuleReference"
+                    && moduleReference.expression.type === "Literal"
+                    && typeof moduleReference.expression.value === "string"
+                ) {
+                    source = moduleReference.expression;
+                }
+            }
+            else {
+                source = node.source;
+            }
+
+            if (source?.value.startsWith(".") && !/\.[cm]?js$/.test(source.value)) {

```

Comment:
Technically this could be failing on a directory module rather than an extensionless filename, but we don't do that anywhere in our codebase, so I'm fine with not checking. Though maybe the rule should have a `DO NOT COPY THIS` comment similar to the declaration bundler.

Reply:
Does that happen even for relative paths? I had figured the startsWith would be enough.

(Trying to not pull in a huge plugin, and a message saying "don't use this would be good yeah)

Reply:
Yes, `import "./foo"` could be referencing a directory with a package.json or just an index file.

Reply:
Ah, I forgot about both cases.